org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/typesystem/computation/CollectionLiteralsTypeComputer.java

```diff
@@ -176,7 +177,17 @@ protected void computeType(XListLiteral literal, JvmGenericType listType, ITypeE
 
 	protected void computeArrayLiteralType(XListLiteral literal, LightweightTypeReference expectedArrayType, ITypeExpectation expectation,
 			ITypeComputationState state) {
-		LightweightTypeReference elementTypeExpectation = expectedArrayType.getComponentType();
+		LightweightTypeReference elementTypeExpectation = null;
```

Comment:
Would not be calling getElementOrComponentType(expectedArrayType, state) enough?

```java
LightweightTypeReference elementTypeExpectation = getElementOrComponentType(expectedArrayType, state);
```

in ElementOrComponentTypeComputer we can override doVisitCompoundTypeReference(CompoundTypeReference):
```java
if (reference.isArray()) {
    for (LightweightTypeReference componentType : reference.getMultiTypeComponents()) {
        if (componentType.isArray()) {
            return componentType.getComponentType();
        }
    }
}
return super.doVisitCompoundTypeReference(reference);
```

Such fix will cover set literals as well.